design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+eventual consistency. 
+
+## Motivation and background
+
+Assume we have a load balancer (LB) and a cluster of three KSQL servers (`A`, `B`, `C`) where `A` 
+is the active server and `B`, `C` are the standbys. `A` receives updates from the source topic 
+partition and writes these updates to its state store and changelog topic. The changelog topic is 
+replicated into the state store of servers `B` and `C`. Note, every Ksql server is the active for 
+some partitions and standby for others. Without loss of generality, we will focus on one partition 
+for now.
+
+Now assume  LB receives a pull query (1) and sends the query to `B` (2). `B` determines that `A` 
+is the active server and forwards the request to `A` (3). `A` executes the query successfully and 
+returns the response.
+
+Failure scenario: Assume that `A` goes down. `B` receives the request, tries to forward it to `A` 
+and fails. What to do now? The current implementation tries to forward the request to `A` 
+(in a busy loop) for a configurable timeout `KSQL_QUERY_PULL_ROUTING_TIMEOUT_MS` . If it has not 
+succeeded in this timeout, the request fails. The next request `B` receives, it will again try to 
+forward it to `A`, since `A` is still the active, and it will fail again. This will happen until 
+rebalancing completes and a new active is elected for the partition. 
+
+There are two steps in the rebalancing procedure that are expensive: 
+  1) One of `B` or `C` is picked as the new active, this takes ~10 seconds based on default configs. 
+  2) Once a new active is chosen, depending on how far behind (lag) it is with respect to the changelog, 
+  it may take up  few seconds or even minutes before the standby has fully caught up to the last committed offset. 
+
+In total, it takes >>10 seconds before a query can start succeeding again. What can we do until a
+new active is ready to serve requests? 
+[KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance) 
+has laid the groundwork to allow us to serve requests from standbys, even if they are still in 
+rebalancing state or have not caught up to the last committed offset of the active. 
+
+Every ksql server now has the information of:
+
+1. Local offset lag: The lag of all partitions the server hosts (whether active or standby)
+2. Routing Metadata for a key: The partition, active host and standy hosts per key
+
+This information allows each server (with some communication, discussed below) to know the global 
+lag information of every topic partition whether it is the standby or active. This enables us to 
+implement a policy where `B` having established that `A` is down (after trying to send a request 
+to `A` and seeing it failed), to decide whether it (`B`) will serve it or `C`. This effectively 
+means that there will be little to no down time in serving requests at the cost of consistency as
```

Comment:
I know other systems might use something like timestamp or operation id of the last read/write and ensure that future reads are at least as current as that timestamp (blocking if necessary for a replica to catch up).  Such a system can ensure consistency without waiting the full cost of a catch up.  That would give a lot of the availability benefits while still ensuring consistency.

Maybe that could be future work if consistency is a focus.

Reply:
Yes, we will add such a configuration parameter.